Add generic-worker test scopes to scm_level_1

petemoore opened the pull request:
@djmitche Not sure if this is a good idea or not.

Basically, sometimes users want to be able to run generic worker (e.g. taskcluster/generic-worker#53), and for this they need the test scopes. Although the test scopes are pretty basic, we probably don't want to give them to everybody. So I decided a reasonable requirement could be that you need to have SCM level 1 as that means you are already marginally vetted. I'm interested to know what you think about this idea - if it is mixing different concerns together, or it is a reasonable fit.

Next, I saw that `mozilla-group:scm_level_1` is auto-generated, so I updated the code to be able to pull in a static list of additional scopes. As I was working on this, it occurred to me that an alternative would just be to reorganise the scopes, e.g. we could grant one additional scope `assume:static:mozilla-group:scm_level_<x>` to `mozilla-group:scm_level_<x>` roles, and then e.g. define role `assume:static:mozilla-group:scm_level_1` to have `assume:project:taskcluster:generic-worker-tester`. But then I thought, this might be overengineering things, especially in light of the fact we intend to implement some kind of parameterized roles.

So this is a halfway house - no additional roles, but changes to "additional static roles" now requiring code changes. That may be a good or bad thing, depending on your point of view!

Lastly, I've little experience with node.js - so if the global var `STATIC_SCOPES` is a bad thing (or anything else I did) or if it should be done more generically across the library somehow (i.e. not just for `make-scm-group-role.js`) then let me know!

Thanks.

CC @mykmelez

petemoore committed:
Add generic-worker test scopes to scm_level_1 - see taskcluster/generic-worker#53

djmitche reviewed (changes requested):
I'm pretty opposed to mixing "scm_level_1" and testing generic-worker -- those are completely unrelated concerns.

The best solution is to not require production scopes for testing.  If that's impossible, then the scopes should be added to a role with the prefix `project:taskcluster:`, and that role should be associated with the necessary roles -- either user roles (e.g., `mozilla-user:mmelez@mozilla.com`), mozillians groups (`mozillians-group:generic-worker-devs`), or perhaps a set of LDAP groups (but I don't think that's appropriate here)

petemoore commented:
> If that's impossible, then the scopes should be added to a role with the prefix `project:taskcluster:`

That is what I did.

> and that role should be associated with the necessary roles -- either user roles (e.g., mozilla-user:mmelez@mozilla.com), mozillians groups (mozillians-group:generic-worker-devs), or perhaps a set of LDAP groups (but I don't think that's appropriate here)

So I think it is reasonable to trust anyone with a mozillians account, since the scopes defined in [`project:taskcluster:generic-worker-tester`](https://tools.taskcluster.net/auth/roles/#project:taskcluster:generic-worker-tester) are quite benign. If we require that you are a member of a mozillians group, that is additional unnecessary overhead for us to administer. We can always change this policy if later we decide it is too loose, but I think it would be inhibitive to not make it easy for people to run generic worker unit tests.

Therefore I've added `assume:project:taskcluster:generic-worker-tester` to `mozillians-user:*`. Note, this is also what we did for granting scopes to run the taskcluster tutorial.

djmitche commented:
Typically we've attached these to the [taskcluster-contributors group](https://tools.taskcluster.net/auth/roles/#mozillians-group:taskcluster-contributors).

petemoore commented:
Ah nice, I hadn't spotted that - I'll do the same! Thanks Dustin.

petemoore commented:
So I've done the following:

* Removed `assume:project:taskcluster:generic-worker-tester` from `mozillians-user:*`
* Moved scopes from `mozillians-group:taskcluster-contributors` into `project:taskcluster:contributor`
* Added `assume:project:taskcluster:generic-worker-tester` to `project:taskcluster:contributor`
* Given `mozillians-group:taskcluster-contributors` the one scope `assume:project:taskcluster:contributor`
* Added `assume:project:taskcluster:contributor` to `mozilla-user:*`

In short, this means that to be a taskcluster contributor, you can either work for Mozilla, or be in the taskcluster-contributors mozillians group (previously you had to be in the mozillians group, which seems overkill if you are already a vouched mozilla employee). In addition, being a taskcluster contributor now also grants you the scopes in `project:taskcluster:generic-worker-tester` which basically allow you to run a test generic worker locally against the queue with a given `provisionerId` / `workerType` combination, and run the integration tests.

djmitche commented:
Repeating from IRC for broader dissemination: "mozilla-user:*" is emphatically not the same as being an employee -- instead, it's anyone with an LDAP account, so all employees, all level-1 users, and any other test accounts, guest accounts, etc.

petemoore commented:
In conclusion, I've removed `assume:project:taskcluster:contributor` from `mozilla-user:*` so the mozillians group membership is the supported mechanism to gain these credentials. Note, this means you'll need group membership before you can even run the worker locally, since it needs to talk to a queue in order to function. But if you are interested in contributing, _hopefully_ setting up a mozillians profile and requesting membership of a group won't stop you.

mykmelez commented:
> But if you are interested in contributing, hopefully setting up a mozillians profile and requesting membership of a group won't stop you.

FWIW, it won't stop me, and I've requested to join taskcluster-contributors (although my reason for running generic-worker locally is to debug test failures in TaskCluster, not to contribute to TaskCluster, so that group's name is a bit misleading for my use case).

Ideally, I'd love to be able to use generic-worker locally without it requiring a queue to function, so I can feed it tasks directly and debug how it runs them. But that's obviously a separate issue.
